examples/discord_bot_with_cogs.py

```diff
@@ -0,0 +1,259 @@
+
+        e.add_field(name="Clan War League Rank",
+                    value=clan.war_league,
+                    inline=False)
+
+        e.add_field(name="Clan Labels",
+                    value="\n".join(label.name for label in clan.labels),
+                    inline=False)
+
+        e.add_field(name="Member Count",
+                    value=f"{clan.member_count}/50",
+                    inline=False)
+
+        e.add_field(
+            name="Clan Record",
+            value=f"Won - {clan.war_wins}\nLost - {clan.war_losses}\n "
+                  f"Draw - {clan.war_ties}",
+            inline=False
+        )
+
+        frame = ""
+        for district in clan.capital_districts:
+            frame += (f"`{f'{district.name}:':<20}` "
+                      f"`{district.hall_level:<15}`\n")
+
+        e2 = discord.Embed(colour=discord.Colour.green(), description=frame,
+                           title="Capital Distracts")
+
+        await ctx.send(embeds=[e, e2])
+
+
+async def send_via_webhook(msg: str) -> None:
+    """Webhooks allow you to send messages to discord without having to
+    have a bot session. Keep in mind that you need to create the webhook
+    URL from Discord"""
+    webhook_url = os.environ.get("DISCORD_WEBHOOK")
+    if webhook_url is None:
+        print("No URL webhook found")
+        return
+
+    async with aiohttp.ClientSession() as session:
+        webhook = discord.Webhook.from_url(os.environ.get("DISCORD_WEBHOOK"),
+                                           session=session)
+        await webhook.send(msg)
+
+
+# The following decorators are listeners. Keep in mind that they must first
+# be registered before they work. The registration process happens during
+# the bots `setup_hook`
+@coc.ClanEvents.member_donations()
+async def on_clan_member_donation(old_member, new_member):
+    final_donated_troops = new_member.donations - old_member.donations
+    msg =(f"{new_member} of {new_member.clan} just donated "
+          f"{final_donated_troops} troops.")
+    print(msg)
+    await send_via_webhook(msg)
+
+
+@coc.ClanEvents.member_received()
+async def on_clan_member_donation_receive(old_member, new_member):
+    final_received_troops = new_member.received - old_member.received
+    msg = (f"{new_member} of {new_member.clan} just received "
+           f"{final_received_troops} troops.")
+    print(msg)
+    await send_via_webhook(msg)
+
+
+@coc.ClanEvents.points()
+async def on_clan_trophy_change(old_clan, new_clan):
+    msg =(f"{new_clan.name} total trophies changed from {old_clan.points} "
+          f"to {new_clan.points}")
+    print(msg)
+    await send_via_webhook(msg)
+
+
+async def main():
+    # Instantiate the client object with the settings you desire
+    coc_client = coc.EventsClient(
+        key_names="coc.py bot example"
+    )
+
+    # Attempt to log into CoC API using your credentials.
+    try:
+        await coc_client.login(os.environ.get("DEV_SITE_EMAIL"),
```

Comment:
I thought one of the reasons for the `setup_hook` function back in the bot subclass is to make a good place to setup async things like these, since it runs once and first. Just wondering what / where the best place to do this is

Reply:
That's definitely a good spot as well. My intention was to not even start the bot if the creds were bad.